src/WebJobs.Script.WebHost/Diagnostics/MetricsEventManager.cs

```diff
@@ -164,24 +166,39 @@ public void Dispose()
             internal void FunctionStarted(FunctionStartedEvent startedEvent)
             {
                 totalExecutionCount++;
-                runningFunctionCount++;
 
                 var metricEventPerFunction = new FunctionMetrics(startedEvent.FunctionMetadata.Name, ExecutionStage.Started, 0);
                 functionMetricsQueue.Enqueue(metricEventPerFunction);
+                var key = GetDictionaryKey(startedEvent.FunctionMetadata.Name, startedEvent.InvocationId);
+                if (!runningFunctions.ContainsKey(key))
+                {
+                    runningFunctions.Add(key, new RunningFunctionInfo(startedEvent.FunctionMetadata.Name, startedEvent.InvocationId, startedEvent.StartTime, startedEvent.Success));
+                }
             }
 
             internal void FunctionCompleted(FunctionStartedEvent completedEvent)
             {
-                if (runningFunctionCount > 0)
-                {
-                    runningFunctionCount--;
-                }
-
                 var functionStage = (completedEvent.Success == false) ? ExecutionStage.Failed : ExecutionStage.Succeeded;
                 long executionTimeInMS = (long)completedEvent.EndTime.Subtract(completedEvent.StartTime).TotalMilliseconds;
 
                 var monitoringEvent = new FunctionMetrics(completedEvent.FunctionMetadata.Name, functionStage, executionTimeInMS);
                 functionMetricsQueue.Enqueue(monitoringEvent);
+                var key = GetDictionaryKey(completedEvent.FunctionMetadata.Name, completedEvent.InvocationId);
+                if (runningFunctions.ContainsKey(key))
+                {
+                    lock (functionMetricEventLockObject)
```

Comment:
Why do you lock here but not above in FunctionStarted? Since keys are unique, you'd never have a situation where you'd get more than one started/completed event for a single invocation ID, right?